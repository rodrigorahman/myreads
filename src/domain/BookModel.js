class BookModel {

    constructor(id, thumbnail, width=128, height=193, title, authors, status, selected=false){
        this.id = id;
        this.thumbnail = thumbnail;
        this.width = width;
        this.height = height;
        this.title = title;
        this.authors = authors;
        this.status = status;
        this.selected = selected;
    }

}

export default BookModel;