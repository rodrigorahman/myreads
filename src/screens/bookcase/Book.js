import React, {Component} from 'react';
import PropTypes from'prop-types';

class Book extends Component {

    static propTypes = {
        book: PropTypes.object.isRequired,
        callbackUpdateState: PropTypes.func.isRequired,
        onAddBooking: PropTypes.func,
        multiSelection: PropTypes.bool.isRequired,
        history: PropTypes.object.isRequired
    };

    addBooking;

    state = {
        cssSelectionGroup: 'book-top-selection-hide'
    };

    bookSelectionHandler = () => {

        if(this.props.multiSelection){
            let cssSelectionGroup = 'book-top-selection';

            let book = this.props.book;

            if (book.selected) {
                cssSelectionGroup = 'book-top-selection-hide';
                book.selected = false;
            } else {
                book.selected = true;
            }

            this.setState({cssSelectionGroup});

            this.props.callbackUpdateState();

        }

    };

    changeOrAddBookcase = (e) => {

        let status = e.target.value;
        this.props.book.status = status;

        if(this.addBooking){
            this.addBooking(this.props.book);
        }
        this.props.callbackUpdateState();
        this.props.history.push('/')
    };

    componentWillMount() {
        this.addBooking = this.props.onAddBooking
    }


    render() {

        const {book} = this.props;
        const {thumbnail, width=128, height=193, title, authors, status} = book;

        return (
            <div className="book">
                <div className="book-top">
                    <div className={this.state.cssSelectionGroup}><input type='radio'></input></div>
                    <div id='img_book' className="book-cover" style={{
                            width: width,
                            height: height,
                            backgroundImage: 'url("'+ thumbnail +'")'
                        }} onDoubleClick={this.bookSelectionHandler}></div>
                    <div className="book-shelf-changer">
                        <select id='sel_change_book' defaultValue={status} onChange={this.changeOrAddBookcase}>
                            <option value="none" disabled>Move to...</option>
                            <option value="currentlyReading">Currently Reading</option>
                            <option value="wantToRead">Want to Read</option>
                            <option value="read">Read</option>
                            <option value="none">None</option>
                        </select>
                    </div>
                </div>
                <div className="book-title">{ title }</div>
                <div className="book-authors">{authors.join(',')}</div>
            </div>
        )
    }
}

export default Book;