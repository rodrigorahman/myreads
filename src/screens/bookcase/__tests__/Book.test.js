import React from 'react';

import {mount, shallow} from 'enzyme';
import Book from "../Book";
import BookModel from "../../../domain/BookModel";

it('renders without crashing with multiSelection', () => {

    let book = {
        id: 1,
        thumbnail: '',
        width: 128,
        height: 193,
        title: '',
        authors: ['autor 1'],
        status: 'none',
        selected: false
    };
    let callbackUpdateState = jest.fn();
    let onAddBooking = jest.fn();
    let multiSelection = true;

    expect(shallow(<Book
        book={book}
        callbackUpdateState={callbackUpdateState}
        onAddBooking={onAddBooking}
        multiSelection={multiSelection}
    />)).toMatchSnapshot();
});


it('change bookstatus with multiSelection and book selected', () => {

    let book = {
        id: 1,
        thumbnail: '',
        width: 128,
        height: 193,
        title: '',
        authors: ['author 1'],
        status: 'none',
        selected: true
    };
    let callbackUpdateState = jest.fn();
    let onAddBooking = jest.fn();
    let multiSelection = true;


    let bookComponent = mount(<Book
        book={book}
        callbackUpdateState={callbackUpdateState}
        onAddBooking={onAddBooking}
        multiSelection={multiSelection}
    />).find('#sel_change_book').simulate('change');

    expect(callbackUpdateState).toHaveBeenCalledTimes(1);

});


it('change bookstatus without multiSelection', () => {

    let book = new BookModel(1,'', 128, 193, '',['author 1'], 'none',true);

    let callbackUpdateState = jest.fn();
    let onAddBooking = jest.fn();
    let multiSelection = true;


    let bookComponent = mount(<Book
        book={book}
        callbackUpdateState={callbackUpdateState}
        onAddBooking={onAddBooking}
        multiSelection={multiSelection}
    />).find('#img_book').simulate('doubleClick');

    expect(callbackUpdateState).toHaveBeenCalledTimes(1);

});

it('double click in image book with selected = false and multiselection=true', () => {

    let book = new BookModel(1,'', 128, 193, '',['author 1'], 'none', false);
    let callbackUpdateState = jest.fn();
    let onAddBooking = jest.fn();
    let multiSelection = true;


    let bookComponent = mount(<Book
        book={book}
        callbackUpdateState={callbackUpdateState}
        onAddBooking={onAddBooking}
        multiSelection={multiSelection}
    />).find('#img_book').simulate('doubleClick');

    expect(callbackUpdateState).toHaveBeenCalledTimes(1);

});