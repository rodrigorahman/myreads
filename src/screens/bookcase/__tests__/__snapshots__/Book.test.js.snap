// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing with multiSelection 1`] = `
<div
  className="book"
>
  <div
    className="book-top"
  >
    <div
      className="book-top-selection-hide"
    >
      <input
        type="radio"
      />
    </div>
    <div
      className="book-cover"
      id="img_book"
      onDoubleClick={[Function]}
      style={
        Object {
          "backgroundImage": "url(\\"\\")",
          "height": 193,
          "width": 128,
        }
      }
    />
    <div
      className="book-shelf-changer"
    >
      <select
        defaultValue="none"
        id="sel_change_book"
        onChange={[Function]}
      >
        <option
          disabled={true}
          value="none"
        >
          Move to...
        </option>
        <option
          value="currentlyReading"
        >
          Currently Reading
        </option>
        <option
          value="wantToRead"
        >
          Want to Read
        </option>
        <option
          value="read"
        >
          Read
        </option>
        <option
          value="none"
        >
          None
        </option>
      </select>
    </div>
  </div>
  <div
    className="book-title"
  />
  <div
    className="book-authors"
  >
    autor 1
  </div>
</div>
`;
